src/geo/projection/globe.js

```diff
@@ -0,0 +1,496 @@
+    const wsRadius = worldSize / (2.0 * Math.PI);
+    const scale = globeECEFUnitsToPixelScale(worldSize);
+
+    if (!offset) {
+        const lat = clamp(tr.center.lat, -MAX_MERCATOR_LATITUDE, MAX_MERCATOR_LATITUDE);
+        const lng = tr.center.lng;
+
+        offset = [
+            mercatorXfromLng(lng) * worldSize,
+            mercatorYfromLat(lat) * worldSize
+        ];
+    }
+
+    // transform the globe from reference coordinate space to world space
+    const posMatrix = mat4.identity(new Float64Array(16));
+    mat4.translate(posMatrix, posMatrix, [offset[0], offset[1], -wsRadius]);
+    mat4.scale(posMatrix, posMatrix, [scale, scale, scale]);
+    mat4.rotateX(posMatrix, posMatrix, degToRad(-tr._center.lat));
+    mat4.rotateY(posMatrix, posMatrix, degToRad(-tr._center.lng));
+
+    return posMatrix;
+}
+
+export function calculateGlobeMercatorMatrix(tr: Transform): mat4 {
+    const worldSize = tr.worldSize;
+    const lat = clamp(tr.center.lat, -MAX_MERCATOR_LATITUDE, MAX_MERCATOR_LATITUDE);
+    const point = new Point(
+        mercatorXfromLng(tr.center.lng) * worldSize,
+        mercatorYfromLat(lat) * worldSize);
+
+    const mercatorZ = mercatorZfromAltitude(1, tr.center.lat) * worldSize;
+    const projectionScaler = mercatorZ / tr.pixelsPerMeter;
+    const zScale = tr.pixelsPerMeter;
+    const ws = worldSize / projectionScaler;
+
+    const posMatrix = mat4.identity(new Float64Array(16));
+    mat4.translate(posMatrix, posMatrix, [point.x, point.y, 0.0]);
+    mat4.scale(posMatrix, posMatrix, [ws, ws, zScale]);
+
+    return posMatrix;
+}
+
+export const GLOBE_ZOOM_THRESHOLD_MIN = 5;
+export const GLOBE_ZOOM_THRESHOLD_MAX = 6;
+
+export function globeToMercatorTransition(zoom: number): number {
+    return smoothstep(GLOBE_ZOOM_THRESHOLD_MIN, GLOBE_ZOOM_THRESHOLD_MAX, zoom);
+}
+
+export function globeBuffersForTileMesh(painter: Painter, tile: Tile, coord: OverscaledTileID, tiles: number): [VertexBuffer, VertexBuffer] {
+    const context = painter.context;
+    const id = coord.canonical;
+    const tr = painter.transform;
+    let gridBuffer = tile.globeGridBuffer;
+    let poleBuffer = tile.globePoleBuffer;
+
+    if (!gridBuffer) {
+        const gridMesh = GlobeSharedBuffers.createGridVertices(id);
+        gridBuffer = tile.globeGridBuffer = context.createVertexBuffer(gridMesh, globeLayoutAttributes, false);
+    }
+
+    if (!poleBuffer) {
+        const poleMesh = GlobeSharedBuffers.createPoleTriangleVertices(tiles, tr.tileSize * tiles, coord.canonical.y === 0);
+        poleBuffer = tile.globePoleBuffer = context.createVertexBuffer(poleMesh, globeLayoutAttributes, false);
+    }
+
+    return [gridBuffer, poleBuffer];
+}
+
+export function globeMatrixForTile(id: CanonicalTileID, globeMatrix: mat4) {
+    const decode = globeDenormalizeECEF(globeTileBounds(id));
+    const posMatrix = mat4.clone(globeMatrix);
```

Comment:
This is potentially dangerous. Clone functions seems to always return 32 bit arrays even if 64 bit ones are passed.

Reply:
ea5cbd4f11c0a02516956a237cbf55df1d8566ba